### PR TITLE
Ensure that speculation analyzer catches replacements which introduces a...

### PR DESCRIPTION
**Pull request by mavasani**
...n error type in any of the ancestor expressions.

Fixes internal bug 1086771

### CODE REVIEW

**Comment by mavasani** [comment]
@srivatsn @shyamnamboodiripad @heejaechang @tmeschter @jmarolf can you please review?

File: src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs [hunk 1]
```diff
@@ -199,6 +199,24 @@ private void EnsureSpeculativeSemanticModel()
 
         #region Semantic comparison helpers
 
+        private bool ReplacementIntroducesErrorType(TExpressionSyntax originalExpression, TExpressionSyntax newExpression)
+        {
+            Debug.Assert(originalExpression != null);
+            Debug.Assert(this.SemanticRootOfOriginalExpression.DescendantNodesAndSelf().Contains(originalExpression));
+            Debug.Assert(newExpression != null);
+            Debug.Assert(this.SemanticRootOfReplacedExpression.DescendantNodesAndSelf().Contains(newExpression));
```
Comment (shyamnamboodiripad) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: [minor] These checks seem to be repeated across bunch of methods below. Consider pulling out into a separate method?

**Comment by shyamnamboodiripad** [comment]
:+1: LGTM

**Comment by shyamnamboodiripad** [comment]
@dotnet-bot test this please

File: src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs [hunk 2]
```diff
@@ -500,7 +518,8 @@ private bool ReplacementChangesSemanticsForNode(TSyntaxNode currentOriginalNode,
             {
                 var originalExpression = (TExpressionSyntax)currentOriginalNode;
                 var newExpression = (TExpressionSyntax)currentReplacedNode;
-                if (!ImplicitConversionsAreCompatible(originalExpression, newExpression))
+                if (!ImplicitConversionsAreCompatible(originalExpression, newExpression) ||
+                    ReplacementIntroducesErrorType(originalExpression, newExpression))
                 {
                     return true;
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: I don't know this code that much but, TypesAreCompatible returning true when it introduced an error feels weird.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs]: ignore my comment, it was due to git diff, code actually belong to different method.

**Comment by heejaechang** [comment]
:+1:

**Pull request merged**